monitor/monitor.go

```diff
@@ -62,55 +77,84 @@ type Monitor struct {
 }
 
 // agentPipeReader reads agent events from the agentPipe and distributes to all listeners
-func (m *Monitor) agentPipeReader(agentPipe io.Reader, stop chan struct{}) {
+func (m *Monitor) agentPipeReader(ctx context.Context, agentPipe io.Reader) {
 	meta, p := payload.Meta{}, payload.Payload{}
 
-	for {
-		select {
-		default:
-			err := payload.ReadMetaPayload(agentPipe, &meta, &p)
-			if err == io.EOF || err == io.ErrUnexpectedEOF {
-				log.Panic("Agent pipe closed, shutting down")
-			} else if err != nil {
-				log.WithError(err).Panic("Unable to read from agent pipe")
-			}
-
-			m.send(&p)
-
-		case <-stop:
-			return
+	for !isCtxDone(ctx) {
+		err := payload.ReadMetaPayload(agentPipe, &meta, &p)
+		if err == io.EOF || err == io.ErrUnexpectedEOF {
+			log.Panic("Agent pipe closed, shutting down")
+		} else if err != nil {
+			log.WithError(err).Panic("Unable to read from agent pipe")
 		}
+
+		m.send(&p)
 	}
 }
 
-// Run starts monitoring.
-func (m *Monitor) Run(npages int, agentPipe io.Reader) {
-	stopAgentPipeReader := make(chan struct{})
-	go m.agentPipeReader(agentPipe, stopAgentPipeReader)
-	defer close(stopAgentPipeReader)
+// Init configures client connection handling and agent event handling.
+// Note that the perf buffer reader is started only when listeners are connected.
+func (m *Monitor) Init(ctx context.Context, npages int, agentPipe io.Reader, server net.Listener) (err error) {
+	// start new listener handler
+	go m.connectionHandler(ctx, server)
+
+	// start agent event pipe reader
+	go m.agentPipeReader(ctx, agentPipe)
 
+	mutex.Lock()
+	defer mutex.Unlock()
+	nPages = npages
+
+	return nil
+}
+
+// startPerfEventReader spawns a singleton goroutine to read and distribute the
+// events. It passes a cancelable context to this goroutine and the cancelFunc
+// is assigned to perfReaderCancel. Note that cancelling parentCtx (e.g. on
+// program shutdown) will also cancel the derived context.
+func (m *Monitor) startPerfEventReader(parentCtx context.Context) {
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	perfEventReaderCtx, cancelFn := context.WithCancel(parentCtx)
+	perfReaderCancel = cancelFn
+
+	go m.perfEventReader(perfEventReaderCtx)
+}
+
+func (m *Monitor) perfEventReader(stopCtx context.Context) {
+	log.Info("Beginning to read perf buffer")
+	defer log.Info("Stopped reading perf buffer")
+
+	// configure BPF perf buffer reader
 	c := bpf.DefaultPerfEventConfig()
-	c.NumPages = npages
+	c.NumPages = nPages
 
-	me, err := bpf.NewPerCpuEvents(c)
+	monEvents, err := bpf.NewPerCpuEvents(c)
 	if err != nil {
-		log.WithError(err).Error("Error while starting monitor")
+		log.WithError(err).Fatal("Cannot initialize BPF perf buffer reader")
```

Comment:
How will this affect the monitor at runtime if this condition is hit? I.e., will the monitor keep restarting over and over? Is this desirable in terms of performance effects? I

Reply:
Good point. We would loop vigorously (https://github.com/cilium/cilium/blob/master/monitor/launch/launcher.go#L77). I'll add a throttle later tonight.